File: boj/분할정복/1074.py
import sys
import math

sys.setrecursionlimit(10000)

def Z(x, y, size):
    half_size = size // 2
    position = int(math.pow(half_size, 2))
    if 0 <= x < half_size and 0 <= y < half_size:
        if size == 2:
            return 0
        else:
            return position * 0 + Z(x, y, half_size)
    elif half_size <= x < size and 0 <= y < half_size:
        if size == 2:
            return 1
        else:
            return position * 1 + Z(x % half_size, y, half_size)
    elif 0 <= x < half_size and half_size <= y < size:
        if size == 2:
            return 2
        else:
            return position * 2 + Z(x, y % half_size, half_size)
    elif half_size <= x < size and half_size <= y < size:
        if size == 2:
            return 3
        else:
            return position * 3 + Z(x % half_size, y % half_size, half_size)
        
    
n, r, c = map(int, sys.stdin.readline().split())
answer = Z(c, r, int(math.pow(2, n)))
print(answer)

File: boj/divide-conquer/1074.js
const fs = require("fs");
//const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");
const input = fs.readFileSync("boj/input.txt").toString().trim().split("\n");

const [N, r, c] = input[0].split(" ").map(Number)

function Z (N, r, c) {
    if (N == 0) return 0;
    const half_size = Math.pow(2, N - 1);
    const numbers = half_size * half_size;
    if (r < half_size && c < half_size) 
        return Z(N - 1, r, c);
    if (r < half_size && c >= half_size)
        return numbers + Z(N - 1, r, c - half_size);
    if (r >= half_size && c < half_size)
        return 2 * numbers + Z(N - 1, r - half_size, c);
    if (r >= half_size && c >= half_size)
        return 3 * numbers + Z(N - 1, r - half_size, c - half_size);
}

console.log(Z(N, r, c))